### ISSUE TITLE
Pluggable transports for iroh-net

### ISSUE THREAD

**Issue by rklaehn** [n0-computer/iroh#2228]
This would be a major enhancement, but it would be very nice if you could have a way to plug in additional transports into iroh-net.

Examples include local transports like BLE, or something like unix datagram sockets or even something like shared memory for communication between local processes.

**Comment by dignifiedquire** [comment]
this is issue is about p2p transports

**Comment by matheus23** [comment]
> For P2P I think iroh-net better focuses on supporting any kinds of IP LANs. Then if someone wants to use BLE or unix socket they only need to turn it into a virtual IP interface. Pluggable transport API can then be something like TUN API.

I disagree. As an end-user application developer, I don't want to force my users to make changes to their ip interfaces to be able to have features like a local BLE-based transport.

**Comment by elsirion** [comment]
Tor support would also be nice, although I fear it would only work for the web socket based alternative protocol (already necessary for web/WASM support afaik).

**Comment by adiibanez** [comment]
Having pluggable transport options would open up interesting possibilities and also provide better integration entry points for community contributions while keeping the integration complexity / cognitive load in check.  

Personally I'm thinking about ble / apple multi peer connectivity and android wifi-direct in particular. 
Maybe also WebTransport for web / wasm support without relay but don't know enough about it nor iroh web / wasm. 
Iroh wasm doesn't have much relevance for me as existing client / server websocket architecture works just fine without requiring relays. 

Apple mpc: https://github.com/adiibanez/iroh_discovery_playground/blob/main/src/lib.rs

According to the objc2 maintainer I'm doing it "completely wrong" at the moment ;) https://github.com/adiibanez/iroh_discovery_playground/blob/main/src/lib.rs

I'll definitely integrate ble and mpc in my projects at some point as a "side, side gig transport" ... but would love to do that within the boundaries of an ongoing iroh integration.

**Comment by rklaehn** [comment]
FYI we are working on a major refactor of iroh networking to use QUIC multipath. Once that is done it would make it easier to add additional transports without having to modify much of iroh.

Basically from the POV of quinn/QUIC each additional transport would be a "special" IPV6 address that gets translated into whatever transport exists underneath.

The only big restriction on the transport would be that it would have to have some way to support the QUIC minimum MTU of 1200 bytes. If your transport does not support such large frames, you would have to work around it somehow.

**Comment by MOZGIII** [comment]
WebTransport builds on top of QUIC, and there is an implementation of it in rust on top of quinn.

From what I've read above (the IPv6 address thing) I don't think it would work right with WebTransport. That said, the main benefit of WebTransport is that it works in the web browsers.

So, to explicitly state what I think here: it would be nice if iroh had proper support for running in the web browser environment via WebTransport, but *that* is unlikely compatible with the idea of IPv6 (maybe on the server side - but definitely not at the clientside); for now, WebTransport in web browsers doesn't do P2P afaik, and it also has a clear separation between the client and server sides baked into the protocol / API - but nonetheless it could work nicely.

Should we, maybe, create a dedicated issue for WebTransport at the web browsers specifically?

UPD: I created #3750

**Comment by teohhanhui** [comment]
> for now, WebTransport in web browsers doesn't do P2P afaik

@kixelated on https://moq.dev/blog/first-cdn/
> I’d also love to get MoQ running over [Iroh](https://www.iroh.computer/) for peer-to-peer action if anybody wants to help.

They're the author of the https://github.com/kixelated/moq library used in Cloudflare's MoQ support: https://developers.cloudflare.com/moq/

**Comment by matheus23** [comment]
Implemented via #3845
